packages/jest-runtime/src/index.ts

```diff
@@ -1164,21 +1173,28 @@ export default class Runtime {
   }
 
   async isolateModulesAsync(fn: () => Promise<void>): Promise<void> {
-    if (this._isolatedModuleRegistry || this._isolatedMockRegistry) {
+    if (
+      this._isolatedModuleRegistry ||
+      this._isolatedMockRegistry ||
+      this._isolatedESMModuleRegistry
+    ) {
       throw new Error(
         'isolateModulesAsync cannot be nested inside another isolateModulesAsync or isolateModules.',
       );
     }
     this._isolatedModuleRegistry = new Map();
     this._isolatedMockRegistry = new Map();
+    this._isolatedESMModuleRegistry = new Map();
     try {
       await fn();
     } finally {
       // might be cleared within the callback
       this._isolatedModuleRegistry?.clear();
       this._isolatedMockRegistry?.clear();
+      this._isolatedESMModuleRegistry?.clear();
       this._isolatedModuleRegistry = null;
       this._isolatedMockRegistry = null;
+      this._isolatedESMModuleRegistry = null;
     }
   }
 
```

Comment:
Could you take a look at `resetModules()` method below (GitHub does not allow to add a comment on that line)? `_isolatedESMModuleRegistry` is not reset. Is that on purpose?

Reply:
Done.